wcfsetup/install/files/acp/update_com.woltlab.wcf_5.4_session_1_cookies.php

```diff
@@ -0,0 +1,103 @@
+<?php
+/**
+ * Sets the new session cookies.
+ * 
+ * @author	Tim Duesterhus
+ * @copyright	2001-2020 WoltLab GmbH
+ * @license	GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package	WoltLabSuite\Core
+ */
+
+use wcf\system\application\ApplicationHandler;
+use wcf\system\form\container\GroupFormElementContainer;
+use wcf\system\form\element\LabelFormElement;
+use wcf\system\form\FormDocument;
+use wcf\system\request\RouteHandler;
+use wcf\system\WCF;
+use wcf\util\CryptoUtil;
+use wcf\util\HeaderUtil;
+
+// 1) Check whether the cookies are already in place.
+$sessionIdOkay = false;
+if (!empty($_COOKIE[COOKIE_PREFIX."acp_session"])) {
+	$cookieValue = CryptoUtil::getValueFromSignedString($_COOKIE[COOKIE_PREFIX."acp_session"]);
+	if ($cookieValue) {
+		$sessionID = \bin2hex($cookieValue);
+		if ($sessionID === WCF::getSession()->sessionID) {
+			$sessionIdOkay = true;
+		}
+	}
+}
+
+$xsrfTokenOkay = false;
+if (!empty($_COOKIE['XSRF-TOKEN'])) {
+	if (CryptoUtil::validateSignedString($_COOKIE['XSRF-TOKEN'])) {
+		$xsrfTokenOkay = true;
+	}
+}
+
+if ($sessionIdOkay && $xsrfTokenOkay) {
+	// The process may continue;
```

Comment:
```suggestion
	// The cookies have already been set in a previous failed upgrade attempt.
```
It took me a moment to realize that this is only a check if the cookies exist because of a previously failed attempt.

Reply:
No, that suggested wording is not correct.

The script will be triggered twice during a single upgrade attempt:

1. The script sets the cookies and hijacks replaces the `SECURITY_TOKEN` variable by injecting the JavaScript code below into the "form response". The newly generated `SECURITY_TOKEN` is designed to be compatible with both the old and the new session logic.
  - The response will render into a dialog containing a short description (a user-friendly wording needs to be added later-on, see the TODO in line 76) and a single button. In the background the JavaScript will run.
  - The User then confirms the dialog by pressing the button.
2. The upgrade routine will trigger the script once again, realizing the the client sent proper cookies, returning here and allowing the upgrade to proceed as usual.

This was the easiest method to replace the SECURITY_TOKEN to be compatible with 5.3, without refactoring everything. In my tests it was super reliable.